spec/controllers/api/orders_controller_spec.rb

```diff
@@ -156,6 +158,102 @@ module Api
+          allow(controller).to receive(:spree_current_user) { regular_user }
+          get :show, id: order.number
+          assert_unauthorized!
+        end
+
+        it "returns the order, as an admin user" do
+          allow(controller).to receive(:spree_current_user) { admin_user }
+          get :show, id: order.number
+          expect_order
+        end
+
+        it "returns the order, as the order distributor owner" do
+          allow(controller).to receive(:spree_current_user) { order.distributor.owner }
+          get :show, id: order.number
+          expect_order
+        end
+
+        it "returns unauthorized, as the order product's supplier owner" do
+          allow(controller).to receive(:spree_current_user) { order.line_items.first.variant.product.supplier.owner }
+          get :show, id: order.number
+          assert_unauthorized!
+        end
+
+        it "returns the order, as the Order Cycle coorinator owner" do
+          allow(controller).to receive(:spree_current_user) { order.order_cycle.coordinator.owner }
+          get :show, id: order.number
+          expect_order
+        end
+      end
+
+      context "as distributor owner" do
+        let!(:order) { create(:completed_order_with_fees, order_cycle: order_cycle, distributor: distributor ) }
+
+        before { allow(controller).to receive(:spree_current_user) { order.distributor.owner } }
+
+        it "can view an order not in a standard state" do
+          order.update_attributes(completed_at: nil, state: 'shipped')
+          get :show, id: order.number
+          expect_order
+        end
+
+        it "returns an order with all required fields" do
+          get :show, id: order.number
+          expect_order
+          expect_detailed_attributes_to_be_present(json_response)
+
+          expect(json_response[:bill_address][:address1]).to eq order.bill_address.address1
+          expect(json_response[:bill_address][:lastname]).to eq order.bill_address.lastname
+          expect(json_response[:ship_address][:address1]).to eq order.ship_address.address1
+          expect(json_response[:ship_address][:lastname]).to eq order.ship_address.lastname
+          expect(json_response[:shipping_method][:name]).to eq order.shipping_method.name
+
+          expect(json_response[:adjustments].first[:label]).to eq "Transaction fee"
+          expect(json_response[:adjustments].first[:amount]).to eq order.adjustments.payment_fee.first.amount.to_s
+          expect(json_response[:adjustments].second[:label]).to eq "Shipping"
+          expect(json_response[:adjustments].second[:amount]).to eq order.adjustments.shipping.first.amount.to_s
+
+          expect(json_response[:payments].first[:amount]).to eq order.payments.first.amount.to_s
+          expect(json_response[:line_items].size).to eq order.line_items.size
+          expect(json_response[:line_items].first[:variant][:product_name]). to eq order.line_items.first.variant.product.name
```

Comment:
I meant something like

```diff
diff --git a/spec/controllers/api/orders_controller_spec.rb b/spec/controllers/api/orders_controller_spec.rb
index 778653663..bf22a3f5b 100644
--- a/spec/controllers/api/orders_controller_spec.rb
+++ b/spec/controllers/api/orders_controller_spec.rb
@@ -220,21 +220,37 @@ module Api
 
         it "returns an order with all required fields" do
           get :show, id: order.number
-          expect_order
-          expect_detailed_attributes_to_be_present(json_response)
 
-          expect(json_response[:bill_address][:address1]).to eq order.bill_address.address1
-          expect(json_response[:bill_address][:lastname]).to eq order.bill_address.lastname
-          expect(json_response[:ship_address][:address1]).to eq order.ship_address.address1
-          expect(json_response[:ship_address][:lastname]).to eq order.ship_address.lastname
+          expect(response.status).to eq 200
+          expect(json_response[:number]).to eq order.number
+
+          expect(json_response.symbolize_keys.keys).to include(*order_detailed_attributes)
+
+          expect(json_response[:bill_address]).to include(
+            'address1' => order.bill_address.address1,
+            'lastname' => order.bill_address.lastname
+          )
+
+          expect(json_response[:ship_address]).to include(
+            'address1' => order.ship_address.address1,
+            'lastname' => order.ship_address.lastname
+          )
+
           expect(json_response[:shipping_method][:name]).to eq order.shipping_method.name
 
-          expect(json_response[:adjustments].first[:label]).to eq "Transaction fee"
-          expect(json_response[:adjustments].first[:amount]).to eq order.adjustments.payment_fee.first.amount.to_s
-          expect(json_response[:adjustments].second[:label]).to eq "Shipping"
-          expect(json_response[:adjustments].second[:amount]).to eq order.adjustments.shipping.first.amount.to_s
+          expect(json_response[:adjustments].first).to include(
+            'label' => "Transaction fee",
+            'amount' => order.adjustments.payment_fee.first.amount.to_s
+          )
+          expect(json_response[:adjustments].second).to include(
+            'label' => "Shipping",
+            'amount' => order.adjustments.shipping.first.amount.to_s
+          )
+
+          expect(json_response[:payments].first).to include(
+            'amount' => order.payments.first.amount.to_s
+          )
 
-          expect(json_response[:payments].first[:amount]).to eq order.payments.first.amount.to_s
           expect(json_response[:line_items].size).to eq order.line_items.size
           expect(json_response[:line_items].first[:variant][:product_name]). to eq order.line_items.first.variant.product.name
         end
@@ -248,10 +264,6 @@ module Api
       def expect_correct_order(json_response, order)
         expect(json_response[:number]).to eq order.number
       end
-
-      def expect_detailed_attributes_to_be_present(json_response)
-        expect(order_detailed_attributes.all?{ |attr| json_response.key? attr.to_s }).to eq true
-      end
     end
 
     private
```

However, I'd question the need to check almost the entire JSON response. That might fit better in a view spec.

Reply:
OMG!!!
expect(json_response.symbolize_keys.keys).to include(*order_detailed_attributes)
I learned 2 new ruby things in one line of code 🎉 
I didnt know symbolize_keys and I didnt know the * (splat) operator :-O
thank you!

yeah, I see your point about tooo many checks for a controller test (but dont want to change it now)... This is replacing the serializer unit tests really...